README.md
# KiTS21

The official repository of the 2021 Kidney and Kidney Tumor Segmentation Challenge

**Current dataset version: `2.2.3` -- Official Frozen Training Set** (see [changelog](changelog.md))

<img src="https://kits21.kits-challenge.org/public/site_media/figures/rendering_dimmed.png" width="400" />

[Challenge Homepage](https://kits21.kits-challenge.org/)

## Timeline

- **Mar 1 - Jul 1**: Annotation, Release, and Refinement of Training Data (*now published!*)
- **July 15**: Further refinement of training set will be complete
- **Aug 23**: Deadline for Intention to Submit & Required Paper (formerly Aug 9)
- **Aug 30 - Sep 13**: Submissions Accepted (formerly Aug 16 - 30)
- **Sep 15**: Results Announced (formerly Sep 1)
- **Sep 27**: Satellite Event at MICCAI 2021

## News

- **July 15, 2021**: The training set has been frozen!
- **July 1, 2021**: The training set has been released! Also we are adding a two-week buffer for final edits to be made based on community feedback, and we are pushing the challenge timeline by two weeks (see above).
- **June 17, 2021**: We've changed the set of classes for the challenge. See [this forum post](https://discourse.kits-challenge.org/t/kits21-challenge-update/354) for details
- **Apr 7, 2021**: We've started using tags and a changelog to keep track of the dataset version
- **Mar 23, 2021**: A draft of the postprocessing code and some preliminary data has been merged into the master branch.
- **Mar 9, 2021**: A preliminary challenge homepage has been published at [kits-challenge.org](https://kits21.kits-challenge.org). You can keep tabs on the data annotation process there.
- **Mar 29, 2020**: A second edition of KiTS was accepted to be held in conjunction with MICCAI 2021 in Strasbourg! More information will be posted here and on the [discussion forum](https://discourse.kits-challenge.org/) when it becomes available.

## Usage

### Installation

1) Install dependency for surface dice:\
`pip install git+https://github.com/JoHof/surface-distance.git` (the original [DeepMind repository](https://github.com/deepmind/surface-distance) is currently not working due to a [missing line comment](https://github.com/deepmind/surface-distance/blob/4315531eb2d449310d47c0850f334cc6a6580543/surface_distance/metrics.py#L102))
2) Clone this repository
3) Install this repository by running `pip install -e .` in the folder where the setup.py file is located

### Download

Start by cloning this repository, but note that **the imaging is not stored here**, it must be downloaded using one of the `get_imaging` scripts in the `starter_code` directory. Currently there are implementations in:

- **python3**: `python3 kits21/starter_code/get_imaging.py`
- **MATLAB**: `matlab kits21/starter_code/get_imaging.m`
- **bash**: `bash kits21/starter_code/get_imaging.sh`
- **julia**: `julia kits21/starter_code/get_imaging.jl`

If you would like to request another implementation of `get_imaging`, please [submit an issue](https://github.com/neheller/kits21/issues/new).

## Folder Structure

### `data/`

```text
kits21
├──data/
|   ├── case_00000/
|   |   ├── raw/
|   |   ├── segmentations/
|   |   ├── imaging.nii.gz
|   |   ├── aggregated_OR_seg.nii.gz
|   |   ├── aggregated_AND_seg.nii.gz
|   |   └── aggregated_MAJ_seg.nii.gz
|   ├── case_00001/
|   |   ├── raw/
|   |   ├── segmentations/
|   |   ├── imaging.nii.gz
|   |   ├── aggregated_OR_seg.nii.gz
|   |   ├── aggregated_AND_seg.nii.gz
|   |   └── aggregated_MAJ_seg.nii.gz
...
|   ├── case_00299/
|   |   ├── raw/
|   |   ├── segmentations/
|   |   ├── imaging.nii.gz
|   |   ├── aggregated_OR_seg.nii.gz
|   |   ├── aggregated_AND_seg.nii.gz
|   |   └── aggregated_MAJ_seg.nii.gz
└── ├── kits.json
```

This is different from [KiTS19](https://github.com/neheller/kits19) because unlike 2019, we now have multiple annotations per "instance" and multiple instances per region.

Consider the "kidney" label in a scan: most patients have two kidneys (i.e., two "instances" of kidney), and each instance was annotated by three independent people. That case's `segmentations/` we would thus have

- `kidney_instance-1_annotation-1.nii.gz`
- `kidney_instance-1_annotation-2.nii.gz`
- `kidney_instance-1_annotation-3.nii.gz`
- `kidney_instance-2_annotation-1.nii.gz`
- `kidney_instance-2_annotation-2.nii.gz`
- `kidney_instance-2_annotation-3.nii.gz`

along with similar collections for `cyst`, and `tumor` regions. The `aggregated_<X>_seg.nii.gz` file is a result of aggregating all of these files by various methods indicated by \<X\>:

- **OR**: A voxel-wise "or" or "union" operator
- **AND**: A voxel-wise "and" or "intersection" operator
- **MAJ**: Voxel-wise majority voting

### `starter_code/`

This folder holds code snippets for viewing and manipulating the data. See [Usage](#Usage) for more information.

### `annotation/`

This folder contains code used to process and import data from the annotation platform. As a participant, there's no reason you should need to run this code, it's only meant to serve as a reference.

## Challenge Information

This challenge will feature significantly more data, several annotations per case, and a number of additional annotated regions. The accepted proposal can be found [on Zenodo](https://doi.org/10.5281/zenodo.3714971), but the most up-to-date information about the challenge can be found on [the KiTS21 homepage](https://kits21.kits-challenge.org/).

## Previous KiTS Challenges

KiTS was first held in conjunction with MICCAI 2019 in Shenzhen. A paper describing that challenge was published in Medical Image Analysis \[[html](https://www.sciencedirect.com/science/article/abs/pii/S1361841520301857)\] \[[pdf](https://arxiv.org/pdf/1912.01054.pdf)\].

```bibtex
@article{heller2020state,
  title={The state of the art in kidney and kidney tumor segmentation in contrast-enhanced CT imaging: Results of the KiTS19 Challenge},
  author={Heller, Nicholas and Isensee, Fabian and Maier-Hein, Klaus H and Hou, Xiaoshuai and Xie, Chunmei and Li, Fengyi and Nan, Yang and Mu, Guangrui and Lin, Zhiyong and Han, Miofei and others},
  journal={Medical Image Analysis},
  pages={101821},
  year={2020},
  publisher={Elsevier}
}
```


changelog.md
# Changelog

This file is meant to keep track of which annotations were changed in each dataset version. Broadly, the *major* version will be incremented with any change to the postprocessing code that causes every region to be regenerated (starting at `v1.x.x`). The *minor* version will be incremented with the addition of new **imaging**, and the *patch* version will be incremented with new **annotations** for the existing imaging.

## [2.2.3] - Removed Duplicate Instance

The cyst region in 205 was an erroneous duplicate of a tumor

## [2.2.2] - Official Dataset Freeze

- Merged in work by Fabian and Dasha on example Docker submissions
- Revisions to
  - `case_00000`
  - `case_00006`
  - `case_00039`
  - `case_00050`
  - `case_00063`
  - `case_00076`
  - `case_00081`
  - `case_00113`
  - `case_00120`
  - `case_00127`
  - `case_00129`
  - `case_00133`
  - `case_00153`
  - `case_00154`
  - `case_00155`
  - `case_00176`
  - `case_00192`
  - `case_00194`
  - `case_00201`
  - `case_00205`
  - `case_00211`
  - `case_00214`
  - `case_00242`
  - `case_00246`
  - `case_00247`
  - `case_00254`
  - `case_00281`

## [2.2.1] - July 1st, 2021: Training Set Release

- Edits to nearly every case

## [2.2] - July 1st, 2021

- Code for evaluation of segmentations has been added. Have a look at kits21/evaluation/readme.md
- Sampling of realistic segmentations from instance annotations
- Code for computing the inter-rater disagreement between sampled segmentations

## [2.1.1] - June 28, 2021

- Complete revisions for
  - `case_00167`
  - `case_00170`
  - `case_00179`
  - `case_00181`
  - `case_00182`
  - `case_00186`
  - `case_00187`
  - cases 194-196
  - cases 199-201
  - cases 203-209
  - `case_00211`
  - `case_00212`
  - cases 216-233
  - `case_00235`
  - `case_00237`
  - `case_00238`
  - cases 240-245
  - `case_00247`
  - `case_00249`
  - `case_00255`
  - cases 258-263
  - `case_00266`
  - `case_00267`
  - `case_00269`
  - `case_00270`
  - `case_00272`
  - cases 274-277
  - `case_00279`
  - `case_00280`
  - cases 283-286
  - `case_00288`
  - `case_00290`
  - `case_00293`
  - `case_00295`
  - `case_00299`
- Incomplete revisions for
  - `case_00168`
  - `case_00169`
  - cases 171-178
  - `case_00180`
  - `case_00184`
  - `case_00185`
  - cases 188-190
  - `case_00193`
  - cases 197-198
  - `case_00202`
  - `case_00210`
  - `case_00213`
  - `case_00214`
  - `case_00234`
  - `case_00236`
  - `case_00239`
  - `case_00246`
  - `case_00248`
  - `case_00278`
  - `case_00281`
  - `case_00282`
  - `case_00287`
  - `case_00289`
  - `case_00291`
  - `case_00292`
  - `case_00294`
  - `case_00296`
  - `case_00297`
- Crowd-workers completed incomplete revisions in cases 0 - 200
  - cases 250-254
  - `case_00256`
  - `case_00257`
  - `case_00264`
  - `case_00265`
  - `case_00268`
  - `case_00271`
  - `case_00273`
  - `case_00278`
  - `case_00281`
  - `case_00282`
  - `case_00287`
  - `case_00289`
  - `case_00291`
  - `case_00292`
  - `case_00294`
  - `case_00296`
  - `case_00297`
- Crowd-workers completed incomplete revisions in cases 0 - 200

## [2.1.0]- June 28, 2021

- Restructured the project as a python package for easier import

## [2.0.6] - June 26, 2021

- Full Annotations for cases 277 - 299
- Complete revisions for
  - cases 146-148
  - cases 151-155
  - cases 159-165`
- Incomplete revisions for
  - `case_00149`
  - `case_00151`
  - `case_00157`
  - `case_00158`
  - `case_00166`

## [2.0.5] - June 25, 2021

- Full Annotations for cases 250 - 276
- Complete revisions for
  - `case_00082`
  - `case_00083`
  - cases 85-88
  - `case_00090`
  - `case_00092`
  - `case_00094`
  - `case_00095`
  - `case_00096`
  - `case_00099`
  - cases 101-107
  - `case_00109`
  - `case_00110`
  - cases 112-116
  - cases 120-122
  - `case_00125`
  - `case_00126`
  - `case_00130`
  - `case_00131`
  - `case_00134`
  - `case_00136`
  - `case_00138`
  - `case_00144`
- Incomplete revisions for
  - `case_00084`
  - `case_00091`
  - `case_00093`
  - `case_00097`
  - `case_00098`
  - `case_00100`
  - `case_00108`
  - `case_00111`
  - cases 117-119
  - `case_00123`
  - `case_00124`
  - `case_00127`
  - `case_00129`
  - `case_00132`
  - `case_00135`
  - `case_00137`
  - cases 139-143
  - `case_00145`

## [2.0.4] - June 24, 2021

- Full annotations for cases 226 - 249
- Complete revisions for
  - `case_00041`
  - `case_00075`
  - `case_00076`
  - `case_00077`
  - `case_00078`
  - `case_00079`
  - `case_00080`
  - `case_00081`

## [2.0.3] - June 23, 2021

- Full annotations for cases 200 - 225
- Incomplete revisions for
  - `case_00017`
  - `case_00019`
  - `case_00021`
  - `case_00023`
  - `case_00026`
  - `case_00033`
  - `case_00051`
  - `case_00056`
  - `case_00057`
  - `case_00058`
  - `case_00060`
  - `case_00065`
  - `case_00066`
  - `case_00068`
  - `case_00071`
  - `case_00150`
- Complete revisions for
  - `case_00016`
  - `case_00018`
  - `case_00020`
  - `case_00022`
  - `case_00024`
  - `case_00025`
  - `case_00027`
  - `case_00028`
  - `case_00029`
  - `case_00030`
  - `case_00031`
  - `case_00032`
  - `case_00034`
  - `case_00035`
  - `case_00036`
  - `case_00038`
  - `case_00039`
  - `case_00042`
  - `case_00050`
  - `case_00052`
  - `case_00053`
  - `case_00054`
  - `case_00055`
  - `case_00059`
  - `case_00061`
  - `case_00062`
  - `case_00064`
  - `case_00067`
  - `case_00069`
  - `case_00070`
  - `case_00072`
  - `case_00073`
  - `case_00074`

## [2.0.2] - June 22, 2021

- Full annotations for cases 150 - 199
- Incomplete revisions for
  - `case_00000`
  - `case_00001`
  - `case_00015`
  - `case_00100`
- Complete revisions for
  - cases 2 - 14
  - `case_00101`
  - `case_00102`

## [2.0.1] - June 18, 2021

- Full annotations for cases 93 - 149

## [2.0.0] - June 17, 2021

- Changed import and aggregation code to ignore ureter, artery, and vein regions
- Deleted existing ureter, artery, and vein instances that were segmented
- New annotations for cases 50 - 92

## [1.0.3] - May 21, 2021

- Full annotations for

kits21/annotation/import.py
import argparse
import shutil
from pathlib import Path
import json

import nibabel as nib
import numpy as np

from kits21.annotation.postprocessing import delineation_to_seg, load_json, write_json
from kits21.configuration.labels import KITS_LABEL_NAMES, LABEL_AGGREGATION_ORDER
from kits21.configuration.paths import SRC_DIR, TRAINING_DIR, TESTING_DIR, CACHE_FILE


def get_case_dir(case):
    assert SRC_DIR is not None, "SRC_DIR was none, this is most likely due to KITS21_SERVER_DATA not being in your " \
                                "environment variables. This functionality is intended to be used only by the " \
                                "KiTS organizers."
    # TODO remove hardcoding -- test both to find it
    page = int(case // 50)
    tst = "training_data"
    if case >= 300:
        tst = "testing_data"
    return (SRC_DIR / tst / "cases_{:05d}".format(page) / "case_{:05d}".format(case)).resolve(strict=True)


def get_all_case_dirs():
    # TODO set this number dynamically
    return [get_case_dir(i) for i in range(400)]


def get_region_dir(case_dir, region):
    return (case_dir / region).resolve(strict=True)


def get_all_region_dirs(case_dir):
    return [r for r in case_dir.glob("*")]


def get_instance_dir(region_dir, instance):
    return (region_dir / "{:02d}".format(instance)).resolve(strict=True)


def get_all_instance_dirs(region_dir):
    return [i for i in region_dir.glob("*")]


def get_existing_instances(region_dir):
    case_id = region_dir.parent.name
    base_dir = Path(__file__).resolve().parent.parent / "data"
    if int(case_id.split("_")[-1]) >= 300:
        base_dir = TESTING_DIR
    seg_dir = base_dir / case_id / "segmentations"
    return [x for x in seg_dir.glob("*{}*".format(region_dir.name))]


def get_delineation(instance_dir, delineation):
    return (instance_dir / "delineation{}".format(delineation)).resolve(strict=True)


def get_all_delineations(instance_dir):
    return [d for d in instance_dir.glob("delineation*")]


def get_most_recent_save(parent_dir):
    # Get latest file and list of remainder
    try:
        srt_files = sorted([s for s in parent_dir.glob("*")])
        latest = srt_files[-1]
    except Exception as e:
        print()
        print("Error finding most recent save in", str(parent_dir))
        raise(e)

    return latest


def update_raw(delineation_path, case_id, in_test_set):
    # Get parent directory (create if necessary)
    destination_parent = TRAINING_DIR / case_id
    if in_test_set:
        destination_parent = TESTING_DIR / case_id
    if not destination_parent.exists():
        destination_parent.mkdir()
    destination_parent = destination_parent / "raw"
    if not destination_parent.exists():
        destination_parent.mkdir()

    custom_hilums = None
    if (destination_parent / "meta.json").exists():
        with (destination_parent / "meta.json").open() as f:
            old_meta = json.loads(f.read())
            if "custom_hilums" in old_meta:
                custom_hilums = old_meta["custom_hilums"]

    # Get source directory
    src = delineation_path.parent.parent.parent.parent

    # Copy all annotation files to destination
    shutil.copytree(str(src), str(destination_parent), dirs_exist_ok=True)

    if custom_hilums is not None:
        with (destination_parent / "meta.json").open() as f:
            new_meta = json.loads(f.read())
        with (destination_parent / "meta.json").open('w') as f:
            new_meta["custom_hilums"] = custom_hilums
            f.write(json.dumps(new_meta, indent=2))


def get_localization(delineation_path):
    return get_most_recent_save(delineation_path.parent.parent / "localization")


def get_artery_localization(delineation_path):
    pth = delineation_path.parent.parent.parent.parent / "artery" / "00" / "localization"
    if not pth.exists():
        return None
    return get_most_recent_save(pth)


def get_image_path(case_id, in_test_set):
    if in_test_set:
        return (TESTING_DIR / case_id / "imaging.nii.gz").resolve(strict=True)
    else:
        return (TRAINING_DIR / case_id / "imaging.nii.gz").resolve(strict=True)


def save_segmentation(case_id, region_type, delineation_path, n1img, in_test_set):
    # Create name of destination file
    annotation_num = int(delineation_path.parent.name[-1])
    instance_num = int(delineation_path.parent.parent.name)
    filename = "{}_instance-{}_annotation-{}.nii.gz".format(region_type, instance_num+1, annotation_num)

    # Get parent directory (create if necessary)
    destination_parent = TRAINING_DIR / case_id
    if in_test_set:
        destination_parent = TESTING_DIR / case_id
    if not destination_parent.exists():
        destination_parent.mkdir()
    destination_parent = destination_parent / "segmentations"
    if not destination_parent.exists():
        destination_parent.mkdir()
    destination = destination_parent / filename

    # Save file
    nib.save(n1img, str(destination))


def run_import(delineation_path):
    # Useful values
    region_type = delineation_path.parent.parent.parent.name
    case_id = delineation_path.parent.parent.parent.parent.name
    in_test_set = False
    if delineation_path.parent.parent.parent.parent.parent.parent.name == "testing_data":
        in_test_set = True

    # Copy updated raw data
    update_raw(delineation_path, case_id, in_test_set)

    # Kidneys require hilum information from the localization
    localization = None
    if region_type == "kidney":
        localization = get_localization(delineation_path)

    # Path to underlying CT scan stored as .nii.gz
    image_path = get_image_path(case_id, in_test_set)

    meta_path = image_path.parent / "raw" / "meta.json"
    meta = load_json(meta_path)

    # Compute and save segmentation based on delineation
    seg_nib = delineation_to_seg(region_type, image_path, delineation_path, meta, localization)
    save_segmentation(case_id, region_type, delineation_path, seg_nib, in_test_set)


def aggregate(parent, region, idnum, agg, affine, agtype="maj"):

    seg_files = [x for x in parent.glob("{}*.nii.gz".format(region))]
    instances = [int(x.stem.split("_")[1].split("-")[1]) for x in seg_files]
    unq_insts = sorted(list(set(instances)))

    reg_agg = None
    for inst in unq_insts:
        inst_agg = None
        n_anns = 0
        for tins, tfnm in zip(instances, seg_files):
            if tins != inst:
                continue
            seg_nib = nib.load(str(tfnm))
            n_anns += 1
            if inst_agg is None:
                inst_agg = np.asanyarray(seg_nib.dataobj)
                affine = seg_nib.affine
            else:
                inst_agg = inst_agg + np.asanyarray(seg_nib.dataobj)

        if agtype == "maj":
            inst = np.greater(inst_agg, n_anns/2).astype(inst_agg.dtype)
        elif agtype == "or":
            inst = np.greater(inst_agg, 0).astype(inst_agg.dtype)
        elif agtype == "and":
            inst = np.equal(inst_agg, n_anns).astype(inst_agg.dtype)

        if reg_agg is None:
            reg_agg = np.copy(inst)
        else:
            reg_agg = np.logical_or(reg_agg, inst).astype(reg_agg.dtype)

    # If no info here, just return what we started with
    if reg_agg is None:
        return agg, affine

    if agg is None:
        agg = idnum*reg_agg
    else:
        agg = np.where(np.logical_not(np.equal(reg_agg, 0)), idnum*reg_agg, agg)

    return agg, affine


def aggregate_case(case_id):
    base_dir = Path(__file__).resolve().parent.parent / "data"
    if int(case_id.split("_")[-1]) >= 300:
        base_dir = TESTING_DIR

    segs =  base_dir / case_id / "segmentations"

    affine = None
    agg = None
    for label_id in LABEL_AGGREGATION_ORDER:
        agg, affine = aggregate(segs, KITS_LABEL_NAMES[label_id], label_id, agg, affine, agtype="or")
    if agg is not None:
        nib.save(
            nib.Nifti1Image(agg.astype(np.int32), affine),
            str(base_dir / case_id / "aggregated_OR_seg.nii.gz")
        )

    affine = None
    agg = None
    for label_id in LABEL_AGGREGATION_ORDER:
        agg, affine = aggregate(segs, KITS_LABEL_NAMES[label_id], label_id, agg, affine, agtype="and")
    if agg is not None:
        nib.save(
            nib.Nifti1Image(agg.astype(np.int32), affine),
            str(base_dir / case_id / "aggregated_AND_seg.nii.gz")
        )

    affine = None
    agg = None
    for label_id in LABEL_AGGREGATION_ORDER:
        agg, affine = aggregate(segs, KITS_LABEL_NAMES[label_id], label_id, agg, affine, agtype="maj")
    if agg is not None:
        nib.save(
            nib.Nifti1Image(agg.astype(np.int32), affine),
            str(base_dir / case_id / "aggregated_MAJ_seg.nii.gz")
        )


def cleanup(case_dir):
    base_dir = Path(__file__).resolve().parent.parent / "data"
    if int(case_dir.name.split("_")[-1]) >= 300:
        base_dir = TESTING_DIR
    case_dir = base_dir / case_dir.name / "raw"
    region_dirs = get_all_region_dirs(case_dir)
    for region_dir in region_dirs:
        instance_dirs = get_all_instance_dirs(region_dir)
        for instance_dir in instance_dirs:
            sessions = [x for x in instance_dir.glob("*")]
            for sess in sessions:
                srt_files = sorted([s for s in sess.glob("*")])
                for f in srt_files[:-1]:
                    f.unlink()


def main(args):
    cache = load_json(CACHE_FILE)
    cli = True
    if args.case is not None:
        case_dirs = [get_case_dir(args.case)]
    else:
        cli = False
        case_dirs = get_all_case_dirs()

    for case_dir in case_dirs:
        print(case_dir.name)
        reaggregate = args.reaggregate
        if cli and args.region is not None:
            region_dirs = [get_region_dir(case_dir, args.region)]
        else:
            cli = False
            region_dirs = get_all_region_dirs(case_dir)

        for region_dir in region_dirs:
            # Skip regions no longer being used
            if region_dir.name in ["artery", "vein", "ureter"]:
                continue
            if cli and args.instance is not None:
                instance_dirs = [get_instance_dir(region_dir, args.instance - 1)]
            else:
                cli = False
                instance_dirs = get_all_instance_dirs(region_dir)

            for instance_dir in instance_dirs:
                if cli and args.delineation is not None:
                    delineations = [get_delineation(instance_dir, args.delineation)]
                else:
                    delineations = get_all_delineations(instance_dir)

                for delineation in delineations:
                    dln_file = get_most_recent_save(delineation)
                    cache_key = str(delineation.relative_to(delineation.parent.parent.parent.parent))
                    if args.regenerate or cache_key not in cache or cache[cache_key] != dln_file.name:
                        run_import(dln_file)
                        cache[cache_key] = dln_file.name
                        write_json(CACHE_FILE, cache)
                        reaggregate = True

            # Delete any instances that were generated before but don't exist anymore
            generated_instances = get_existing_instances(region_dir)
            for gi in generated_instances:
                if int(gi.stem.split("instance-")[1][0]) not in [int(x.name)+1 for x in instance_dirs]:
                    print("Deleting legacy file:", str(gi.name))
                    gi.unlink()
                    reaggregate = True

        if reaggregate:
            aggregate_case(case_dir.name, )
        
        # Clean up all unused raw files
        cleanup(case_dir)


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("-c", "--case", help="The index of the case to import", type=int)
    parser.add_argument("-r", "--region", help="The type of region to import", type=str)
    parser.add_argument("-i", "--instance", help="The index of the instance of that region to import", type=int)
    parser.add_argument("-d", "--delineation", help="The index of the delineation of that instance to import (1, 2, or 3)", type=int)
    parser.add_argument("--regenerate", help="Regenerate segmentations regardless of cached values", action="store_true")
    parser.add_argument("--reaggregate", help="Reaggregate segmentations regardless of whether it was changed", action="store_true")
    if __name__ == "__main__":
        cl_args = parser.parse_args()
        main(cl_args)
